### PR TITLE
[GNNFlux] Fix `Temporal graph classification` tutorial

### PR DESCRIPTION
**Pull request by aurorarossi**
#545

### CODE REVIEW

File: GNNGraphs/src/temporalsnapshotsgnngraph.jl [hunk 1]
```diff
@@ -54,10 +54,10 @@ GNNGraph:
...
 struct TemporalSnapshotsGNNGraph{G<:GNNGraph, D<:DataStore}
-    num_nodes::Vector{Int}   
-    num_edges::Vector{Int}
+    num_nodes::AbstractVector{Int}   
+    num_edges::AbstractVector{Int}
     num_snapshots::Int
-    snapshots::Vector{G}
+    snapshots::AbstractVector{G}
     tgdata::D   
 end
```
Comment (CarloLucibello) [inline GNNGraphs/src/temporalsnapshotsgnngraph.jl:62]: these changes shouldn't be needed because of this
https://github.com/JuliaGraphs/GraphNeuralNetworks.jl/pull/573/files#diff-1aa1e7fe1f4b8ca3c78522bb95d8600bab501a27d9f49bf9fe56df14aaa901aaR65

try after a rebase

File: GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+The TemporalBrains dataset contains a collection of functional brain connectivity networks from 1000 subjects obtained from resting-state functional MRI data from the [Human Connectome Project (HCP)](https://www.humanconnectome.org/study/hcp-young-adult/document/extensively-processed-fmri-data-documentation).
+Functional connectivity is defined as the temporal dependence of neuronal activation patterns of anatomically separated brain regions.
+
+The graph nodes represent brain regions and their number is fixed at 102 for each of the 27 snapshots, while the edges, representing functional connectivity, change over time.
+For each snapshot, the feature of a node represents the average activation of the node during that snapshot.
+Each temporal graph has a label representing gender ('M' for male and 'F' for female) and age group (22-25, 26-30, 31-35, and 36+).
+The network's edge weights are binarized, and the threshold is set to 0.6 by default.
+
+````julia
+brain_dataset = TemporalBrains()
```
Comment (CarloLucibello) [inline GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md:37]: not necessarily blocking this PR, but we should add an option to the `TemporalBrains` constructor to download a subset of the dataset, so that we don't need to download gigabytes of data to run a tutorial.

File: GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+dataset TemporalBrains:
+  graphs  =>    1000-element Vector{MLDatasets.TemporalSnapshotsGraph}
+````
+
+After loading the dataset from the MLDatasets.jl package, we see that there are 1000 graphs and we need to convert them to the `TemporalSnapshotsGNNGraph` format.
+So we create a function called `data_loader` that implements the latter and splits the dataset into the training set that will be used to train the model and the test set that will be used to test the performance of the model.
+
+````julia
+function data_loader(brain_dataset)
+	graphs = brain_dataset.graphs
```
Comment (CarloLucibello) [inline GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md:50]: indentation

File: GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    dataset = Vector{TemporalSnapshotsGNNGraph}(undef, length(graphs))
+    for i in 1:length(graphs)
+        graph = graphs[i]
+        dataset[i] = TemporalSnapshotsGNNGraph(GNNGraphs.mlgraph2gnngraph.(graph.snapshots))
+		# Add graph and node features
+        for t in 1:27
+			s = dataset[i].snapshots[t]
```
Comment (CarloLucibello) [inline GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md:57]: indentation

File: GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            s.ndata.x = [I(102); s.ndata.x']
+        end
+        dataset[i].tgdata.g = Float32.(Flux.onehot(graph.graph_data.g, ["F", "M"]))
+    end
+    # Split the dataset into a 80% training set and a 20% test set
+    train_loader = dataset[1:200]
+    test_loader = dataset[201:250]
+    return train_loader, test_loader
```
Comment (CarloLucibello) [inline GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md:65]: not cleat why it is using only 250 graphs

File: GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+We now implement a simple model that takes a `TemporalSnapshotsGNNGraph` as input.
+It consists of a `GINConv` applied independently to each snapshot, a `GlobalPool` to get an embedding for each snapshot, a pooling on the time dimension to get an embedding for the whole temporal graph, and finally a `Dense` layer.
+
+First, we start by adapting the `GlobalPool` to the `TemporalSnapshotsGNNGraphs`.
+
+````julia
+function (l::GlobalPool)(g::TemporalSnapshotsGNNGraph, x::AbstractVector)
+    h = [reduce_nodes(l.aggr, g[i], x[i]) for i in 1:(g.num_snapshots)]
+    sze = size(h[1])
+    reshape(reduce(hcat, h), sze[1], length(h))
```
Comment (CarloLucibello) [inline GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md:90]:
```suggestion
    z = reduce(hcat, h)
    return reshape(z, :, length(h))
```

File: GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    gin = GINConv(mlp, 0.5)
+    globalpool = GlobalPool(mean)
+    dense = Dense(nhidden => 2)
+    return GenderPredictionModel(gin, mlp, globalpool, dense)
+end
+
+function (m::GenderPredictionModel)(g::TemporalSnapshotsGNNGraph)
+    h = m.gin(g, g.ndata.x)
+    h = m.globalpool(g, h)
+    h = mean(h, dims=2)
```
Comment (CarloLucibello) [inline GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md:117]: instead of doing the hcat, we can do directly the `mean` in the globalpool.

File: GraphNeuralNetworks/docs/src_tutorials/introductory_tutorials/temporal_graph_classification.jl [hunk 8]
```diff
@@ -143,8 +147,8 @@ function train(dataset)
 end
 
 
-train(brain_dataset)
+train(brain_dataset);
```
Comment (CarloLucibello) [inline GraphNeuralNetworks/docs/src_tutorials/introductory_tutorials/temporal_graph_classification.jl:151]: missing closing ticks?

File: GNNGraphs/src/temporalsnapshotsgnngraph.jl (hunk 1 above)
Comment (aurorarossi) [reply GNNGraphs/src/temporalsnapshotsgnngraph.jl:62]: Yes it works

File: GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md (hunk 5 above)
Comment (aurorarossi) [reply GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md:65]: Because of the time needed to train 1000 is too high

Comment (CarloLucibello) [reply GraphNeuralNetworks/docs/src/tutorials/temporal_graph_classification.md:61]: This should be commented in the tutorial

**Review by CarloLucibello — APPROVED** [review]
